iree/hal/driver_registry.h

```diff
@@ -102,6 +102,17 @@ typedef struct iree_hal_driver_registry_s iree_hal_driver_registry_t;
 IREE_API_EXPORT iree_hal_driver_registry_t* IREE_API_CALL
 iree_hal_driver_registry_default();
 
+// Creates and initializes an empty device driver registry
+//
+// Thread-safe.
+IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_driver_registry_create(
+    iree_allocator_t allocator, iree_hal_driver_registry_t** out_registry);
+
+// Unregisters each factory and frees the associated registry resources.
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_hal_driver_registry_unregister_and_free(
```

Comment:
the allocator passed to _create should be stored on the registry such that this does not need it to be passed in